### PR TITLE
Fix a memleak in cos_linker

### PR DESCRIPTION
**Pull request by RobertGiff**
### Summary of this Pull Request (PR)

cos_linker devoured memory, this PR helps to alleviate the memory usage.

### Intent for your PR

Choose one (Mandatory):

- [ ] This PR is for a code-review and is intended to get feedback, but not to be pulled yet.
- [X] This PR is mature, and ready to be integrated into the repo.

### Reviewers (Mandatory):

@Others, @gparmer 

### Code Quality

As part of this pull request, I've considered the following:

[Style](https://github.com/gparmer/composite/raw/ppos/doc/style_guide/composite_coding_style.pdf):

- [X] Comments adhere to the Style Guide (SG)
- [X] Spacing adhere's to the SG
- [X] Naming adhere's to the SG
- [X] All other aspects of the SG are adhered to, or exceptions are justified in this pull request
- [ ] I have run the auto formatter on my code before submitting this PR (see doc/auto_formatter.md for instructions)

[Code Craftsmanship](http://www2.seas.gwu.edu/~gparmer/posts/2016-03-07-code-craftsmanship.html):

- [X] I've made an attempt to remove all redundant code
- [X] I've considered ways in which my changes might impact existing code, and cleaned it up
- [X] I've formatted the code in an effort to make it easier to read (proper error handling, function use, etc...)
- [X] I've commented appropriately where code is tricky
- [X] I agree that there is no "throw-away" code, and that code in this PR is of high quality

### Testing

I've tested the code using the following test programs (provide list here):

- llbooter_test.sh
- llbooter_pong.sh

### CODE REVIEW

File: src/platform/linker/loadall.c [hunk 1]
```diff
@@ -20,27 +20,53 @@
...
+	char name[MAX_SYMB_LEN];
+	unsigned long vma_val;
+};
+
 unsigned long getsym(bfd *obj, char* symbol)
 {
         long storage_needed;
-        asymbol **symbol_table;
+        asymbol **symbol_table = NULL;
+	static struct private_symtab *private_symtabs;
```
Comment (gparmer) [inline src/platform/linker/loadall.c]: Inconsistent use of spaces/tabs.  This file is incorrectly spaces, but you should at least be consistent with yourself.

File: src/platform/linker/loadall.c [hunk 2]
```diff
@@ -20,27 +20,53 @@
...
+		        printl(PRINT_HIGH, "no symbols in object file\n");
+		        exit(-1);
+		}
+
+		symbol_table = (asymbol **) malloc (storage_needed);
+		assert(symbol_table);
+		number_of_symbols = bfd_canonicalize_symtab(obj, symbol_table);
+		private_symtabs = malloc (number_of_symbols * sizeof(struct private_symtab));
+		assert(private_symtabs);
+		int i;
```
Comment (gparmer) [inline src/platform/linker/loadall.c]: style

File: src/platform/linker/loadall.c (hunk 1 above)
Comment (RobertGiff) [reply src/platform/linker/loadall.c]: Oops, thank you. I'll update the file to use tabs throughout.

File: src/platform/linker/loadall.c [hunk 3]
```diff
@@ -20,27 +20,53 @@
...
+struct private_symtab{
+	char name[MAX_SYMB_LEN];
+	unsigned long vma_val;
+};
+
 unsigned long getsym(bfd *obj, char* symbol)
 {
         long storage_needed;
-        asymbol **symbol_table;
+        asymbol **symbol_table = NULL;
```
Comment (Others) [inline src/platform/linker/loadall.c]: This optimization disallows any freeing of symbols returned by this function. I assume the linker never does this. Is there no way for the linker just to free unused symbols?

File: src/platform/linker/loadall.c (hunk 1 above)
Comment (Others) [inline src/platform/linker/loadall.c]: Would there be any benifit to caching more than the just the previous call?

File: src/platform/linker/cl_macros.h [hunk 4]
```diff
@@ -38,6 +38,7 @@
 #define MAX_SYMBOLS (1024 * 64)
 #define MAX_TRUSTED 32
 #define MAX_SYMB_LEN 256
+#define MAX_FILE_NAME_LEN 500
```
Comment (Others) [inline src/platform/linker/cl_macros.h]: `512`?

File: src/platform/linker/loadall.c [hunk 5]
```diff
@@ -20,27 +20,53 @@
 #include <assert.h>
 #include <sys/mman.h>
 
+char *service_name = NULL;
+
+struct private_symtab{
```
Comment (Others) [inline src/platform/linker/loadall.c]: Space after `private_symtab`

File: src/platform/linker/loadall.c [hunk 6]
```diff
@@ -20,27 +20,53 @@
...
+	}
+
+	assert(private_symtabs != NULL);
 
         //notes: symbol_table[i]->flags & (BSF_FUNCTION | BSF_GLOBAL)
         for (i = 0; i < number_of_symbols; i++) {
-                if(!strcmp(symbol, symbol_table[i]->name)){
-                        return symbol_table[i]->section->vma + symbol_table[i]->value;
+                if(!strcmp(symbol,  private_symtabs[i].name)){
+                        return private_symtabs[i].vma_val;
```
Comment (Others) [inline src/platform/linker/loadall.c]: Why can't we just free `private_symtabs` after this?

File: src/platform/linker/loadall.c (hunk 3 above)
Comment (RobertGiff) [reply src/platform/linker/loadall.c]: I'm not sure. However as @gparmer is rewriting the linker, I don't suppose implementing this is worth it.

File: src/platform/linker/loadall.c (hunk 6 above)
Comment (RobertGiff) [reply src/platform/linker/loadall.c]: Yes and no. We should be able to free private_symtabs, but not here. If we free here we would possibly crash next time this function is executed if it expected the cached private_symtabs to be there. We could cache vma_val, but that is adding complexity through additional variables. I added a free just prior to allocating the next private_symtabs and tested with llbooter_pong.sh and it worked fine. Can't test as of right now with rumpkernboot.sh, though.

**Pull request merged**